File: Support Vector Regression (SVR)/SVR/svr1.py
# -*- coding: utf-8 -*-
"""
Created on Sun Dec 24 12:42:25 2017

@author: Ali Baloch
"""

# Support Vector Regression SVR

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
# Just to check weather the employee is telling truth about his/her salary or bluffing
X = dataset.iloc[:, 1:2].values # transforming the column into matrix of feature so that we can apply polynomial rules
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X = sc_X.fit_transform(X)
y = np.ravel(sc_y.fit_transform(y.reshape(-1, 1)))  


# Fitting Regression Model to the Dataset
from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
regressor.fit(X, y)

#predict a new result with Ploynimial Regression
# inversing the tranformation to get the real result
y_pred = sc_y.inverse_transform(regressor.predict(sc_X.transform(np.array([[6.5]])))) # array of one line and one column: due to double brackets
# inversing the tranformation to get the real result

# Visualising the SVR results

plt.scatter(X, y, color = 'black')
plt.plot(X, regressor.predict(X), color = 'blue')
plt.title("Truth or Bluff (SVR)")
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()
# outlier detected for CEO


#Visualising the SVR result for higher resolution and smoother curve
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid),1))
plt.scatter(X, y, color = 'black')
plt.plot(X_grid, regressor.predict(X_grid), color = 'purple')
plt.title("Truth or Bluff (SVR)")
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

File: Support Vector Regression (SVR)/SVR/svr1.R
# SVR 
# Data Preprocessing 

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3] # select only level and salary column

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
#library(caTools)
#set.seed(123)
#split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
#training_set = subset(dataset, split == TRUE)
#test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)


# Fitting the SVR to dataset
#install.packages('e1071')
library(e1071)
regressor = svm(formula = Salary ~ .,
                data = dataset,
                type = 'eps-regression')

# predict the new result with Regression
y_pred = predict(regressor, data.frame(Level= 6.5))


# Visualising the  Regression result
library(ggplot2)
#x_grid = seq(min(dataset$Level), max(dataset$Level),0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color ='red') + 
  geom_line(aes(x = dataset$Level, y = predict(regressor, newdata = dataset)),
            color ='blue') +
  ggtitle('Truth or Bluff (SVR Model)') +
  xlab('Levels') +
  ylab('Salary')
# CEO is detect as outlier and SVR didn't consider it as it is very far from other datatypes



# Preidict the new result with Regression
y_pred_poly = predict(regressor, data.frame(Level =6.5, 
                                            Level2 =6.5^2, 
                                            Level3 = 6.5^3, 
                                            Level4 = 6.5^4))
